js-envs/src/main/scala/org/scalajs/jsenv/RunConfig.scala

```diff
@@ -0,0 +1,157 @@
+  private[jsenv] val failValidationForTest: Boolean
+) {
+  import RunConfig.OnOutputStream
+
+  private def this() = {
+    this(
+        onOutputStream = None,
+        inheritOutput = true,
+        inheritError = true,
+        logger = NullLogger,
+        failValidationForTest = false)
+  }
+
+  def withOnOutputStream(onOutputStream: OnOutputStream): RunConfig =
+    copy(onOutputStream = Some(onOutputStream))
+
+  def withInheritOut(inheritOutput: Boolean): RunConfig =
+    copy(inheritOutput = inheritOutput)
+
+  def withInheritErr(inheritError: Boolean): RunConfig =
+    copy(inheritError = inheritError)
+
+  def withLogger(logger: Logger): RunConfig =
+    copy(logger = logger)
+
+  private[jsenv] def withFailValidationForTest(failValidationForTest: Boolean): RunConfig =
+    copy(failValidationForTest = failValidationForTest)
+
+  private def copy(onOutputStream: Option[OnOutputStream] = onOutputStream,
+      inheritOutput: Boolean = inheritOutput,
+      inheritError: Boolean = inheritError,
+      logger: Logger = logger,
+      failValidationForTest: Boolean = failValidationForTest): RunConfig = {
+    new RunConfig(onOutputStream, inheritOutput, inheritError, logger,
+        failValidationForTest)
+  }
+
+  /** Validates constraints on the config itself. */
+  private def validate(): Unit = {
+    if (onOutputStream.isEmpty && (!inheritOutput || !inheritError)) {
+      throw new IllegalArgumentException("You may not set inheritOutput or " +
+          "inheritError to false without setting onOutputStream.")
+    }
+
+    if (failValidationForTest)
+      throw new RunConfig.FailedValidationForTestException
+  }
+
+  /** Creates a new [[Validator]] for this [[RunConfig]].
+   *
+   *  A [[RunConfig]] passed to a [[JSEnv]] '''must''' be validated via this method.
+   *
+   *  @see [[Validator]] for details.
+   */
+  def validator(): Validator = new Validator()
+
+  /** Support validator for [[RunConfig]].
+   *
+   *  Validators allow us to add options to [[RunConfig]] in a forward compatible manner. Every [[JSEnv]] must
+   *
+   *  1. create a [[Validator]] (through [[RunConfig#validator]]).
+   *  1. inform it of the [[JSEnv]]'s capabilities (through calls to `supportsX()`).
+   *  1. invoke [[validate]] to enusre that all set config options are handled.
+   */
+  final class Validator private (
+      inheritIO: Boolean,
+      onOutputStream: Boolean
+  ) {
+    def this() = this(false, false)
+
+    /** The caller supports [[inheritOutput]] and [[inheritError]]. */
+    def supportsInheritIO(): Validator = copy(inheritIO = true)
+
+    /** The caller supports [[onOutputStream]]. */
+    def supportsOnOutputStream(): Validator = copy(onOutputStream = true)
+
+    /** Validates that the bound [[RunConfig]] only sets options that are supported.
+     *
+     *  @throws java.lang.IllegalArgumentException if there are unsupported options.
+     */
+    def validate(): Unit = {
+      val cfg = RunConfig.this
```

Comment:
IMO a more idiomatic API would be that `Validator` is not tight to a `RunConfig` upfront, but is declared in the `object RunConfig` instead and that `validate()` takes the `config: RunConfig` as a formal parameter.

This avoids the `#` types, and allows a `JSEnv` to create its `RunConfig.Validator` once and for all, and reuse it to validate all the `RunConfig`s it receives.

Reply:
Good point.